lib/utils.py

```diff
@@ -305,6 +305,8 @@ def extract_schema_from_snapshot(snapshot_path):
 
     :returns: (2, 3, 1)
     """
+    if not os.path.exists(snapshot_path):
+        raise(OSError)
```

Comment:
It is verified in lib/options.py::check() before running any test:

> $ ./test/test-run.py --snapshot xxx
>
> Path /home/alex/projects/tarantool-meta/r/t-6/xxx does not exist

Even if we'll remove the check (let's assume that we unintentionally broke test-run or that there is a race condition and the file can be removed during the testing), we'll get the following:

> <..backtrace..>
> [001] FileNotFoundError: [Errno 2] No such file or directory: '/home/alex/projects/tarantool-meta/r/t-6/xxx'

I guess you want this check, because of the unit test, but it worth to mention it in the commit message. At least one may understand this commit as fix of the actual test-run behaviour.

I removed all beforehand checks to hit into the new one and look how it'll be shown in the test-run output:

> <..backtrace..>
> [001] OSError

The error message is not descriptive. In fact, it raises a class instead of its instance. Proper way to raise the exception would look so (see [SO](https://stackoverflow.com/a/36077407/1598057) for the explanation):

```python
# <..somewhere at top of the file..>
if PY2:
    FileNotFoundError = IOError
# <...>
    if not os.path.exists(snapshot_path):
        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), snapshot_path)
```

I also was wondered, why it is not raised by some `open(snapshot_path, 'rb')` call deeper by the code and I see: we read the file via tarantoolctl in `xlog_rows()`. So the check should be in this function: we can use it more in a future (I do that in my [WIP patch](https://github.com/tarantool/test-run/commit/5fbe2195d4d1d662d8553af668a91acc12483a5f) for example).

Reply:
Update exception to `FileNotFoundError`

Reply:
I still think that we should perform the check in the same function, where Popen() is called, (it is `xlog_rows()`), and still think that a reason of the change should be mentioned in the commit message.

Reply:
Could you explain why we should place code to check the existence of a snapshot path to `xlog_rows()` if we can check it earlier in `extract_schema_from_snapshot()`?

Popen can raise exception `FileNotFound` only when the command not found:
```
--- a/lib/utils.py
+++ b/lib/utils.py
@@ -287,7 +287,7 @@ def xlog_rows(xlog_path):
 
         Assume tarantool and tarantoolctl is in PATH.
     """
-    cmd = ['tarantoolctl', 'cat', xlog_path, '--format=json', '--show-system']
+    cmd = ['tarantoolctl1', 'cat', xlog_path, '--format=json', '--show-system']
     with open(os.devnull, 'w') as devnull:
         process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=devnull)
     for line in process.stdout.readlines():
```
```
======================================================================
ERROR: test_extract_schema_from_snapshot (test_lib_utils.TestUtils)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sergeyb/sources/MRG/tarantool/test-run/test/unittest/test_lib_utils.py", line 12, in test_extract_schema_from_snapshot
    v = utils.extract_schema_from_snapshot(snapshot_path)
  File "/home/sergeyb/sources/MRG/tarantool/test-run/lib/utils.py", line 315, in extract_schema_from_snapshot
    for row in xlog_rows(snapshot_path):
  File "/home/sergeyb/sources/MRG/tarantool/test-run/lib/utils.py", line 292, in xlog_rows
    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=devnull)
  File "/usr/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1702, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'tarantoolctl1'

----------------------------------------------------------------------
Ran 2 tests in 0.010s

FAILED (errors=1)
```


And Popen() cannot detect that path passed as a parameter doesn't exist. 
```
--- a/lib/utils.py
+++ b/lib/utils.py
@@ -287,7 +287,7 @@ def xlog_rows(xlog_path):
 
         Assume tarantool and tarantoolctl is in PATH.
     """
-    cmd = ['tarantoolctl', 'cat', xlog_path, '--format=json', '--show-system']
+    cmd = ['tarantoolctl', 'cat', 'path', '--format=json', '--show-system']
     with open(os.devnull, 'w') as devnull:
         process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=devnull)
     for line in process.stdout.readlines():
```
```
======================================================================
FAIL: test_extract_schema_from_snapshot (test_lib_utils.TestUtils)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sergeyb/sources/MRG/tarantool/test-run/test/unittest/test_lib_utils.py", line 13, in test_extract_schema_from_snapshot
    self.assertEqual(v, (2, 3, 1))
AssertionError: None != (2, 3, 1)

----------------------------------------------------------------------
Ran 2 tests in 0.027s

FAILED (failures=1)
make: *** [Makefile:21: test_unittest] Error 1
```

So what is a reason to move condition with `if os.path.exists()` to a `xlog_rows()`?

Reply:
> And Popen() cannot detect that path passed as a parameter doesn't exist.

That's my key point. If we'll imagine that `xlog_rows()` uses more typical way to read a file (using built-in `open()`), it would return an error on its own. It is meaningful: this function reads the file, so it raises an error, when the file is not exists. It looks logical to lean on this in `extract_schema_from_snapshot()`: usage of Popen() is the implementation detail of `xlog_rows()`.

It would be more important, if we would use `xlog_rows()` from several places.

Anyway, that's not critical and I don't insist.

Reply:
Moved to `xlog_rows()`:

```python
--- a/lib/utils.py
+++ b/lib/utils.py
@@ -1,3 +1,4 @@
+import errno
 import os
 import sys
 import collections
@@ -24,6 +25,9 @@ UNIX_SOCKET_LEN_LIMIT = 107
 PY3 = sys.version_info[0] == 3
 PY2 = sys.version_info[0] == 2
 
+if PY2:
+    FileNotFoundError = IOError
+
 if PY3:
     string_types = str,
     integer_types = int,
@@ -283,6 +287,8 @@ def xlog_rows(xlog_path):
 
         Assume tarantool and tarantoolctl is in PATH.
     """
+    if not os.path.exists(xlog_path):
+        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), xlog_path)
     cmd = ['tarantoolctl', 'cat', xlog_path, '--format=json', '--show-system']
     with open(os.devnull, 'w') as devnull:
         process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=devnull)
```